core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -385,277 +387,49 @@ public void run() {
-              listener.closed(reason, new Metadata());
-            }
-          });
-        }
-      }
     }
 
     @Override
-    public void writeMessage(InputStream message) {
-      if (realStream == null) {
-        synchronized (this) {
-          if (realStream == null) {
-            pendingMessages.add(new PendingMessage(message, messageCompressionEnabled));
-            return;
-          }
-        }
+    public void onSuccess(ClientTransport transport) {
+      if (transport == null) {
+        stream.maybeClosePrematurely(Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+        return;
       }
-      realStream.writeMessage(message);
-    }
-
-    @Override
-    public void flush() {
-      if (realStream == null) {
-        synchronized (this) {
-          if (realStream == null) {
-            pendingFlush = true;
-            return;
-          }
-        }
+      if (!updateTimeoutHeader(callOptions.getDeadlineNanoTime(), headers)) {
+        stream.maybeClosePrematurely(Status.DEADLINE_EXCEEDED);
+        return;
       }
-      realStream.flush();
+      stream.createStream(transport.newStream(method, headers, listener));
     }
 
     @Override
-    public void cancel(Status reason) {
-      maybeClosePrematurely(reason);
-      realStream.cancel(reason);
-    }
-
-    @Override
-    public void halfClose() {
-      if (realStream == null) {
-        synchronized (this) {
-          if (realStream == null) {
-            pendingHalfClose = true;
-            return;
-          }
-        }
-      }
-      realStream.halfClose();
-    }
-
-    @Override
-    public void request(int numMessages) {
-      if (realStream == null) {
-        synchronized (this) {
-          if (realStream == null) {
-            pendingFlowControlRequests += numMessages;
-            return;
-          }
-        }
-      }
-      realStream.request(numMessages);
-    }
-
-    @Override
-    public synchronized void setCompressor(Compressor c) {
-      compressor = c;
-      if (realStream != null) {
-        realStream.setCompressor(c);
-      }
-    }
-
-    @Override
-    public synchronized void setDecompressionRegistry(DecompressorRegistry registry) {
-      this.decompressionRegistry = registry;
-      if (realStream != null) {
-        realStream.setDecompressionRegistry(registry);
-      }
-    }
-
-    @Override
-    public boolean isReady() {
-      if (realStream == null) {
-        synchronized (this) {
-          if (realStream == null) {
-            return false;
-          }
-        }
-      }
-      return realStream.isReady();
-    }
-
-    @Override
-    public synchronized void setMessageCompression(boolean enable) {
-      if (realStream != null) {
-        realStream.setMessageCompression(enable);
-      } else {
-        messageCompressionEnabled = enable;
+    public void onFailure(Throwable t) {
+      stream.maybeClosePrematurely(Status.fromThrowable(t));
+      if (t instanceof InterruptedException) {
```

Comment:
This isn't necessary and is misleading. There isn't a problem with interruption for the callback, because that is a detail when using `Future.get()`, which we aren't doing here (yes, it is being done, but it isn't our responsibility to handle, just like we don't expect `ExecutionException`).